src/runtime_src/core/pcie/linux/device_linux.cpp

```diff
@@ -59,13 +64,66 @@ get_render_value(const std::string dir)
     return instance_num;
 }
 
-
 inline pdev
 get_pcidev(const xrt_core::device* device)
 {
   return pcidev::get_dev(device->get_device_id(), device->is_userpf());
 }
 
+
+static std::vector<uint64_t> 
+get_counter_status_from_sysfs(const std::string &mon_name_address, 
+                              const std::string &sysfs_file_name, 
+                              size_t size, 
+                              const xrt_core::device* device)
+{
+  auto pdev = get_pcidev(device);
+
+  /* Get full path to "name" sysfs file. 
+   * Then use that path to form full path to "counters"/"status" sysfs file 
+   * which contains counter/status data for the monitor
+   */
+  std::string name_path = pdev->get_sysfs_path(mon_name_address, "name");
+
+  std::size_t pos = name_path.find_last_of('/');
+  if (std::string::npos == pos) {
+    std::string msg = "Invalid path for name sysfs node for " + mon_name_address;
+    throw xrt_core::query::sysfs_error(msg);
+  }
+
+  std::string path = name_path.substr(0, pos+1);
+  path += sysfs_file_name;
+
+  std::vector<uint64_t> val_buf(size);
+
+  std::ifstream ifs(path);
+  if (!ifs) {
+    std::string msg = "Incomplete counter data in " + path + ". Using default values in results.";
+    xrt_core::message::send(xrt_core::message::severity_level::info, "XRT", msg);
+    return val_buf;
+  }
+
+  std::string buffer;
+  ifs >> buffer;
+
+  size_t idx = 0;
+  // Each line in counter/status sysfs node contains string for integer value (unsigned long long) 
+  while (!ifs.eof() && idx < size) {
+    val_buf[idx] = strtoull(buffer.c_str(), nullptr, 10);
+    idx++;
+    buffer = "";
+    ifs >> buffer;
+  }
+
+  if (idx < size) {
+    std::string msg = "Incomplete counter data in " + path + ". Using default values in results.";
+    xrt_core::message::send(xrt_core::message::severity_level::info, "XRT", msg);
+  }
```

Comment:
OK, but if you want to read integers, why not reading them at the first place?
Think Python on non speed critical code. :-)
```suggestion
try {
  // Enable exception on reading error
  ifs.exceptions(std::ifstream::failbit);
  for (std::size_t idx = 0; idx < size; ++idx)
    ifs >> val_buf[idx];
  } catch (const std::ios_base::failure& fail) {
  std::string msg = "Incomplete data for counter " + idx " in " + path + " because " + fail.what()
    + "\nUsing 0 in results starting at position " + idx + ".";
  xrt_core::message::send(xrt_core::message::severity_level::info, "XRT", msg);
}
```
I let you adapt the code to compile, perhaps adding some `to_string` and so on.

Reply:
The current code works fine. Do you see any blocking issue with the current code?

Reply:
Sure, sure...
Code quality? Beauty? Empathy with the reader? :-)

Reply:
Ronan's suggestion for reading the buffer is great.  Much clearer to me what the code is actually trying to do.
I agree this about code quality, beautiful code, and readability.

Reply:
I have updated this block of code as per your suggestion about using try-catch for exception. However, I modified the message to
std::string msg = "Incomplete counter data read from " + path + " due to " + fail.what()
+ ".\n Using 0 as default value in results.";
I think mentioning "idx" in the message won't be very helpful to the user as user does not have any knowledge about the
std::vector used in implementation.